test/testSwCanvas.cpp

```diff
@@ -94,3 +94,251 @@ TEST_CASE("Memory Pool", "[tvgSwCanvas]")
+    REQUIRE(canvas->draw() == Result::Success);
+    REQUIRE(canvas->update() == Result::InsufficientCondition);
+    REQUIRE(canvas->clear() == Result::InsufficientCondition);
+    REQUIRE(canvas->sync() == Result::Success);
+
+    REQUIRE(canvas->clear() == Result::Success);
+
+    //Invalid shape
+    REQUIRE(canvas->update(ptr) == Result::InsufficientCondition);
+
+    REQUIRE(Initializer::term() == Result::Success);
+}
+
+TEST_CASE("Synchronized Drawing", "[tvgSwCanvas]")
+{
+    REQUIRE(Initializer::init(0) == Result::Success);
+
+    auto canvas = SwCanvas::gen();
+    REQUIRE(canvas);
+
+    REQUIRE(canvas->sync() == Result::InsufficientCondition);
+    REQUIRE(canvas->draw() == Result::InsufficientCondition);
+
+    uint32_t buffer[100*100];
+    REQUIRE(canvas->target(buffer, 100, 100, 100, SwCanvas::Colorspace::ARGB8888) == Result::Success);
+
+    REQUIRE(canvas->draw() == Result::InsufficientCondition);
+    REQUIRE(canvas->sync() == Result::InsufficientCondition);
+
+    //Invalid Shape
+    auto shape = Shape::gen();
+    REQUIRE(shape);
+    REQUIRE(canvas->push(std::move(shape)) == Result::Success);
+
+    REQUIRE(canvas->draw() == Result::Success);
+    REQUIRE(canvas->sync() == Result::Success);
+    REQUIRE(canvas->clear() == Result::Success);
+
+    auto shape2 = Shape::gen();
+    REQUIRE(shape2);
+    REQUIRE(shape2->appendRect(0, 0, 100, 100) == Result::Success);
+    REQUIRE(shape2->fill(255, 255, 255, 255) == Result::Success);
+
+    REQUIRE(canvas->push(std::move(shape2)) == Result::Success);
+    REQUIRE(canvas->draw() == Result::Success);
+    REQUIRE(canvas->sync() == Result::Success);
+
+    REQUIRE(Initializer::term() == Result::Success);
+}
+
+TEST_CASE("Asynchronized Drawing", "[tvgSwCanvas]")
+{
+    //Use multi-threading
+    REQUIRE(Initializer::init(2) == Result::Success);
+
+    auto canvas = SwCanvas::gen();
+    REQUIRE(canvas);
+
+    uint32_t buffer[100*100];
+    REQUIRE(canvas->target(buffer, 100, 100, 100, SwCanvas::Colorspace::ARGB8888) == Result::Success);
+
+    for (int i = 0; i < 3; ++i) {
+        auto shape = Shape::gen();
+        REQUIRE(shape);
+        REQUIRE(shape->appendRect(0, 0, 100, 100) == Result::Success);
+        REQUIRE(shape->fill(255, 255, 255, 255) == Result::Success);
+
+        REQUIRE(canvas->push(std::move(shape)) == Result::Success);
+    }
+
+    REQUIRE(canvas->draw() == Result::Success);
+    REQUIRE(canvas->sync() == Result::Success);
+
+    REQUIRE(Initializer::term() == Result::Success);
+}
+
+TEST_CASE("Viewport", "[tvgSwCanvas]")
+{
+    REQUIRE(Initializer::init(0) == Result::Success);
+
+    auto canvas = SwCanvas::gen();
+    REQUIRE(canvas);
+
+    REQUIRE(canvas->viewport(25, 25, 100, 100) == Result::Success);
+
+    uint32_t buffer[100*100];
+    REQUIRE(canvas->target(buffer, 100, 100, 100, SwCanvas::Colorspace::ARGB8888) == Result::Success);
+
+    REQUIRE(canvas->viewport(25, 25, 50, 50) == Result::Success);
```

Comment:
Does this code affect the actual rendering results?
For the results below, the final vport result values are 0, 0, 100, 00, and I wanted the actual output to be displayed in the upper left corner.
The output result is at the bottom right.
If the viewport has changed, shouldn't the update be performed internally? 
Because that's how push() works.

```diff
diff --git a/examples/Svg.cpp b/examples/Svg.cpp
index 259fd01f..b9b8ed79 100644
--- a/examples/Svg.cpp
+++ b/examples/Svg.cpp
@@ -120,7 +120,11 @@ void tvgSwTest(uint32_t* buffer)

 void drawSwView(void* data, Eo* obj)
 {
+    swCanvas->viewport(WIDTH - 100, HEIGHT - 100, 100, 100);
+    swCanvas->update();
+    swCanvas->viewport(0, 0, 100, 100);
     if (swCanvas->draw() == tvg::Result::Success) {
+    swCanvas->viewport(WIDTH - 100, 0, 100, 100);
         swCanvas->sync();
     }
 }

```

Reply:
oh... I mean. It was a call between update() and draw(). 
My explain about `REQUIRE(canvas->viewport(25, 25, 50, 50) == Result::Success);` is wrong.

To summarize again..., `viewport()` called after calling `update()`(or `push()`) is not actually applied and returns Success.
Since this means that our API is affected by the order, my opinion was that, like the push() API, the update() function should be called again internally when the viewport is updated.
I think, if someone who does not know this, sets the viewport between update() and draw()(or after update())and performs a transition, An unintended area may be displayed in the first frame.

This is my test code
```diff
diff --git a/examples/Svg.cpp b/examples/Svg.cpp
index 259fd01f..f792030a 100644
--- a/examples/Svg.cpp
+++ b/examples/Svg.cpp
@@ -120,7 +120,11 @@ void tvgSwTest(uint32_t* buffer)

 void drawSwView(void* data, Eo* obj)
 {
+    printf("1) viewport(924 924 100 100) => %d\n", static_cast<int>(swCanvas->viewport(WIDTH - 100, HEIGHT - 100, 100, 100)));
+    swCanvas->update();
+    printf("2) viewport(0, 0, 100, 100) => %d\n", static_cast<int>(swCanvas->viewport(0, 0, 100, 100)));
     if (swCanvas->draw() == tvg::Result::Success) {
+    printf("3) viewport(WIDTH - 100, 0, 100, 100) => %d\n", static_cast<int>(swCanvas->viewport(WIDTH - 100, 0, 100, 100)));
         swCanvas->sync();
     }
 }
diff --git a/src/renderer/tvgCanvas.h b/src/renderer/tvgCanvas.h
index 7e2b7c3f..1a81431d 100644
--- a/src/renderer/tvgCanvas.h
+++ b/src/renderer/tvgCanvas.h
@@ -98,6 +98,7 @@ struct Canvas::Impl
         auto flag = RenderUpdateFlag::None;
         if (refresh || force) flag = RenderUpdateFlag::All;

+printf("in update function call viewport(vport) : %d %d %d %d\n", vport.x, vport.y, vport.w, vport.h);
         renderer->viewport(vport);

         if (paint) {
@@ -113,6 +114,7 @@ struct Canvas::Impl

     Result draw()
     {
+printf("in draw function call viewport(vport) : %d %d %d %d\n", vport.x, vport.y, vport.w, vport.h);
         renderer->viewport(vport);

         if (drawing || paints.empty() || !renderer->preRender()) return Result::InsufficientCondition;
```

result
```
$ ./build/examples/Svg
...
in update function call viewport(vport) : 0 0 2147483647 2147483647
1) viewport(924 924 100 100) => 0 // < - Success
in update function call viewport(vport) : 924 924 100 100
2) viewport(0, 0, 100, 100) => 0  // <- Success
in draw function call viewport(vport) : 0 0 100 100
3) viewport(WIDTH - 100, 0, 100, 100) => 2 // <- InsufficientCondition
```
The actual results are rendered in 924 924 100 100.